crates/ruff_python_formatter/src/expression/expr_bin_op.rs

```diff
@@ -1,12 +1,215 @@
+};
 
 #[derive(Default)]
-pub struct FormatExprBinOp;
+pub struct FormatExprBinOp {
+    parentheses: Option<Parentheses>,
+}
+
+impl FormatRuleWithOptions<ExprBinOp, PyFormatContext<'_>> for FormatExprBinOp {
+    type Options = Option<Parentheses>;
+
+    fn with_options(mut self, options: Self::Options) -> Self {
+        self.parentheses = options;
+        self
+    }
+}
 
 impl FormatNodeRule<ExprBinOp> for FormatExprBinOp {
     fn fmt_fields(&self, item: &ExprBinOp, f: &mut PyFormatter) -> FormatResult<()> {
-        write!(f, [verbatim_text(item.range)])
+        let ExprBinOp {
+            left,
+            right,
+            op,
+            range: _,
+        } = item;
+
+        let should_break_right = self.parentheses == Some(Parentheses::Custom);
+
+        if should_break_right {
+            let left = left.format().memoized();
+            let right = right.format().memoized();
+
+            write!(
+                f,
+                [best_fitting![
+                    // The whole expression on a single line
+                    format_args![left, space(), op.format(), space(), right],
+                    // Break the right, but keep the left flat
+                    format_args![
+                        left,
+                        text(" "),
+                        op.format(),
+                        text(" "),
+                        group(&right).should_expand(true),
+                    ],
+                    // Break after the operator, try to keep the right flat, otherwise expand it
+                    format_args![
+                        text("("),
+                        block_indent(&format_args![
+                            left,
+                            hard_line_break(),
+                            op.format(),
+                            space(),
+                            group(&right),
+                        ]),
+                        text(")")
+                    ],
+                ]]
+            )
+        } else {
+            let comments = f.context().comments().clone();
+            let operator_comments = comments.dangling_comments(item.as_any_node_ref());
+            let needs_space = !is_simple_power_expression(item);
+
+            let before_operator_space = if needs_space {
+                soft_line_break_or_space()
+            } else {
+                soft_line_break()
+            };
+
+            write!(
+                f,
+                [
+                    left.format(),
+                    before_operator_space,
+                    op.format(),
+                    trailing_comments(operator_comments),
+                ]
+            )?;
+
+            // Format the operator on its own line if it has any trailing comments at the right side has leading comments.
+            if !operator_comments.is_empty() && comments.has_leading_comments(right.as_ref().into())
```

Comment:
nit: you can treat this similar to `before_operator_space` so it becomes one write statement